### PR TITLE
ed, routing, jormungandr et al: from shapes.txt to GeoJson

### PR DESCRIPTION
**Pull request by TeXitoi**
From reading a shapes.txt with the gtfs connector to giving a beautiful GeoJson in a public transport section.

Fix http://jira.canaltp.fr/browse/NAVITIAII-410

### CODE REVIEW

File: source/jormungandr/tests/routing_tests.py [hunk 1]
```diff
@@ -294,3 +294,26 @@ def test_max_attemps_debug(self):
...
+         - len of geojson/coordinates is 3 (and thus,
+           stop_date_times is not used to create the geojson)
+        """
+        response = self.query_region(journey_basic_query, display=False)
+        print response['journeys'][0]['sections'][1]
+        assert(len(response['journeys']) == 2)
+        assert(len(response['journeys'][0]['sections']) == 3)
+        assert(response['journeys'][0]['sections'][1]['type'] == 'public_transport')
+        assert(len(response['journeys'][0]['sections'][1]['stop_date_times']) == 2)
+        assert(len(response['journeys'][0]['sections'][1]['geojson']['coordinates']) == 3)
```
Comment (antoine-de) [inline source/jormungandr/tests/routing_tests.py:318]: won't it be nice to check that it's a valid geojson ?

File: source/ed/connectors/gtfs_parser.cpp [hunk 2]
```diff
@@ -813,6 +848,17 @@ void TripsGtfsHandler::handle_line(Data& data, const csv_row& row, bool) {
...
+    // get shape if possible
+    nt::LineString shape;
+    try {
+        const auto& multi_shape = gtfs_data.shapes.at(row.at(shape_id_c));
+        if (!multi_shape.empty()) { shape = multi_shape.front(); }
+        if (multi_shape.size() > 1) {
+            LOG4CPLUS_WARN(logger, "shape of trip " << row[trip_c]
+                           << " is a multi line, taking only the first line.");
+        }
+    } catch (...) {}
```
Comment (antoine-de) [inline source/ed/connectors/gtfs_parser.cpp]: roh it's possible to catch narrower exception no ? (or no exception at all there)
if you test the shape_id_c after that you just have to test the existence in the shapes map

**Comment by koch-t** [comment]
Shapes.txt has the column shape_dist_traveled, combined with the one in stop_times.txt, it's the most accurate way of determining the start and end of a shape for a journey especially if it's complex shape such as a loop. It's  however an optional field and thus not always filled in

File: source/sql/alembic/versions/2a8c7b6401e1_add_shape_column_to_vehicle_journey.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+revision = '2a8c7b6401e1'
+down_revision = '52b017632678'
+
+from alembic import op
+import sqlalchemy as sa
+import geoalchemy2 as ga
+
+
+def upgrade():
+    ### commands auto generated by Alembic - please adjust! ###
```
Comment (antoine-de) [inline source/sql/alembic/versions/2a8c7b6401e1_add_shape_column_to_vehicle_journey.py]: I think you might remove this comment (same for downgrade)

**Comment by TeXitoi** [comment]
Modified version to do all the work in gtfs2ed. It will allow to use the shape_dist_traveled or any other exotic method to give a shape to 2 successive stop times.

It seems to work great except it's quite costy on big data (5328 shapes and 10797168 stop points, it takes 10GB of RAM and 25 minutes).

**Comment by TeXitoi** [comment]
Jenkins, test this please

**Comment by TeXitoi** [comment]
I still have to test GeoJson

File: source/ed/ed_reader.cpp [hunk 4]
```diff
@@ -496,14 +497,15 @@ void EdReader::fill_journey_patterns(nt::Data& data, pqxx::work& work){
...
+        "stop_point_id, journey_pattern_id, ST_AsText(shape_from_prev) as shape "
+        "FROM navitia.journey_pattern_point";
 
     pqxx::result result = work.exec(request);
     for(auto const_it = result.begin(); const_it != result.end(); ++const_it){
         nt::JourneyPatternPoint* jpp = new nt::JourneyPatternPoint();
 
         const_it["uri"].to(jpp->uri);
-        const_it["\"order\""].to(jpp->order);
+        const_it["order"].to(jpp->order);
```
Comment (TeXitoi) [inline source/ed/ed_reader.cpp]: this diff must be removed

File: source/routing/tests/routing_api_test_data.h [hunk 5]
```diff
@@ -338,7 +338,10 @@ struct routing_api_data {
         b.sa("stopA", A.lon(), A.lat());
         b.sa("stopB", B.lon(), B.lat());
         //we add a very fast bus (2 seconds) to be faster than walking and biking
-        b.vj("A")("stop_point:stopB", 8*3600 + 1*60, 8*3600 + 1 * 60)("stop_point:stopA", 8*3600 + 1 * 60 + 2 ,8*3600 + 1*60 + 2);
+        b.vj("A")
+            ("stop_point:stopB", 8*3600 + 1*60, 8*3600 + 1 * 60)
+            ("stop_point:stopA", 8*3600 + 1 * 60 + 2 ,8*3600 + 1*60 + 2)
+            .st_shape({B, I, A});
```
Comment (antoine-de) [inline source/routing/tests/routing_api_test_data.h:344]: shouldn't you also test the shape in the routing_api_test ?

File: source/jormungandr/tests/routing_tests.py (hunk 1 above)
Comment (TeXitoi) [inline source/jormungandr/tests/routing_tests.py:318]: done

File: source/routing/tests/routing_api_test_data.h (hunk 5 above)
Comment (TeXitoi) [inline source/routing/tests/routing_api_test_data.h:344]: Done.

**Comment by TeXitoi** [comment]
Jenkins, test this please

**Pull request merged**